### PR TITLE
[BugFix] Remove invalid predicates after mv rewrite

### LINKED ISSUE #36536: inconsistent results  after mv rewrite for left outer join
results is inconsistent  after rewrite by mv.

### Steps to reproduce the behavior (Required)
1. CREATE TABLE 
```
 CREATE TABLE `test10` (
  `event_id` int(11) NULL COMMENT "Unique query id",
  `event_type` varchar(65533) NULL COMMENT "Unique query id",
  `event_time` datetime NULL COMMENT "Query start time"
) ENGINE=OLAP
DUPLICATE KEY(`event_id`, `event_type`)
COMMENT "OLAP"
PARTITION BY RANGE(`event_time`)
(PARTITION p20230105 VALUES [("2023-01-05 00:00:00"), ("2023-01-06 00:00:00")),
PARTITION p20230106 VALUES [("2023-01-06 00:00:00"), ("2023-01-07 00:00:00")))
DISTRIBUTED BY HASH(`event_id`) BUCKETS 16
PROPERTIES (
"replication_num" = "1",
"dynamic_partition.enable" = "false",
"dynamic_partition.time_unit" = "DAY",
"dynamic_partition.time_zone" = "Asia/Shanghai",
"dynamic_partition.start" = "-3",
"dynamic_partition.end" = "3",
"dynamic_partition.prefix" = "p",
"dynamic_partition.buckets" = "1",
"dynamic_partition.history_partition_num" = "0",
"in_memory" = "false",
"enable_persistent_index" = "false",
"replicated_storage" = "true",
"fast_schema_evolution" = "true",
"compression" = "LZ4"
);


 CREATE TABLE `test11` (
  `event_id1` int(11) NULL COMMENT "Unique query id",
  `event_type1` varchar(65533) NULL COMMENT "Unique query id",
  `event_time1` datetime NULL COMMENT "Query start time"
) ENGINE=OLAP
DUPLICATE KEY(`event_id1`, `event_type1`)
COMMENT "OLAP"
PARTITION BY RANGE(`event_time1`)
(PARTITION p20230105 VALUES [("2023-01-05 00:00:00"), ("2023-01-06 00:00:00")),
PARTITION p20230106 VALUES [("2023-01-06 00:00:00"), ("2023-01-07 00:00:00")))
DISTRIBUTED BY HASH(`event_id1`) BUCKETS 16
PROPERTIES (
"replication_num" = "1",
"dynamic_partition.enable" = "false",
"dynamic_partition.time_unit" = "DAY",
"dynamic_partition.time_zone" = "Asia/Shanghai",
"dynamic_partition.start" = "-3",
"dynamic_partition.end" = "3",
"dynamic_partition.prefix" = "p",
"dynamic_partition.buckets" = "1",
"dynamic_partition.history_partition_num" = "0",
"in_memory" = "false",
"enable_persistent_index" = "false",
"replicated_storage" = "true",
"fast_schema_evolution" = "true",
"compression" = "LZ4"
);


CREATE MATERIALIZED VIEW `mv1` (`event_id`, `event_time`, `event_time1`)
PARTITION BY (`event_time`)
DISTRIBUTED BY HASH(`event_id`) BUCKETS 1
REFRESH ASYNC
PROPERTIES (
"replicated_storage" = "true",
"partition_refresh_number" = "2",
"force_external_table_query_rewrite" = "CHECKED",
"replication_num" = "1",
"storage_medium" = "HDD"
)
AS SELECT `a`.`event_id`, `a`.`event_time`, `b`.`event_time1`
FROM `test`.`test10` AS `a` LEFT OUTER JOIN `test`.`test11` AS `b` ON (`a`.`event_id` = `b`.`event_id1`) AND (`a`.`event_time` = `b`.`event_time1`);

```
3. INSERT INTO 

```
insert into test10 values(1, 'a', '2023-01-05 10:20:22');
insert into test10 values(2, 'b', '2023-01-05 10:20:22');
insert into test10 values(11, 'aa', '2023-01-06 10:20:22');
insert into test10 values(22, 'bb', '2023-01-06 10:20:22');
insert into test11 values(1, 'a', '2023-01-05 10:20:22');
insert into test11 values(2, 'b', '2023-01-05 10:20:23');
insert into test11 values(11, 'aa', '2023-01-06 10:20:22');
insert into test11 values(22, 'bbx', '2023-01-06 10:20:22');
```
5. SELECT  

```

MySQL [test]>  select * from (select event_id, event_time, event_time1 from test10 a left outer join test11 b on a.event_id = b.event_id1 and a.event_time = b.event_time1 ) xx where xx.event_time >= "2023-01-05 00:00:00" and xx.event_time < "2023-01-06 00:00:00";
+----------+---------------------+---------------------+
| event_id | event_time          | event_time1         |
+----------+---------------------+---------------------+
|        1 | 2023-01-05 10:20:22 | 2023-01-05 10:20:22 |
+----------+---------------------+---------------------+
1 row in set (0.02 sec)

MySQL [test]> set enable_materialized_view_rewrite = false;
Query OK, 0 rows affected (0.01 sec)

MySQL [test]>  select * from (select event_id, event_time, event_time1 from test10 a left outer join test11 b on a.event_id = b.event_id1 and a.event_time = b.event_time1 ) xx where xx.event_time >= "2023-01-05 00:00:00" and xx.event_time < "2023-01-06 00:00:00";
+----------+---------------------+---------------------+
| event_id | event_time          | event_time1         |
+----------+---------------------+---------------------+
|        1 | 2023-01-05 10:20:22 | 2023-01-05 10:20:22 |
|        2 | 2023-01-05 10:20:22 | NULL                |
+----------+---------------------+---------------------+
2 rows in set (0.03 sec)

```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `main-7802670`

### PR DESCRIPTION
**Pull request by ABingHuang**
Why I'm doing:
For queries has pushdown derived predicates, the rewritten plan may be wrong, which will lead to invalid results. Originally, the isPushdown and isRedundant fields of ScalarOperator may be incorrectly set.
What I'm doing:
fix isPushdown and isRedundant after predicates pushdown rules.

Fixes #36536 

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarRangePredicateExtractor.java [hunk 1]
```diff
@@ -119,9 +126,7 @@ private ScalarOperator rewrite(ScalarOperator predicate, boolean onlyExtractColu
             if (!checkStatisticsEstimateValid(extractExpr)) {
                 extractExpr.setNotEvalEstimate(true);
             }
-            // TODO: merge `setFromPredicateRangeDerive` into `setRedundant`
-            result.forEach(f -> f.setRedundant(true));
-            extractExpr.setRedundant(true);
+            result.forEach(f -> optimizerContext.addPushdownDerivedPredicate(f));
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarRangePredicateExtractor.java:129]: `setRedundant ` is used not only for mv rewrite, is this change safe?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 2]
```diff
@@ -566,10 +566,12 @@ public OptExpression rewrite() {
         final ReplaceColumnRefRewriter mvColumnRefRewriter =
                 MvUtils.getReplaceColumnRefWriter(mvExpression, mvColumnRefFactory);
 
-        final List<ScalarOperator> mvConjuncts = MvUtils.getAllValidPredicates(mvExpression);
+        final List<ScalarOperator> mvConjuncts = MvUtils.getAllValidPredicates(mvExpression, optimizerContext);
         ScalarOperator mvPartitionCompensate = compensateMVPartitionPredicate(mvConjuncts, mvColumnRefRewriter);
         if (mvPartitionCompensate != ConstantOperator.TRUE) {
-            mvConjuncts.addAll(MvUtils.getAllValidPredicates(mvPartitionCompensate));
+            Utils.extractConjuncts(mvPartitionCompensate).stream()
+                    .filter(p -> !optimizerContext.getPushdownDerivedPredicates().contains(p))
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:573]: Why not push this into `getAllValidPredicates ` method?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/JoinPredicatePushdown.java [hunk 3]
```diff
@@ -403,38 +407,38 @@ private void getPushdownPredicatesFromEquivalenceDerive(
         if (join.getJoinType().isLeftSemiJoin()) {
             for (ScalarOperator p : derivedPredicates) {
                 if (rightOutputColumns.containsAll(p.getUsedColumns())) {
-                    p.setIsPushdown(true);
+                    optimizerContext.addPushdownDerivedPredicate(p);
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/JoinPredicatePushdown.java:410]: What's the difference between `p.setIsPushdown(true);` and `optimizerContext.addPushdownDerivedPredicate(p);`?

**Review by satanson — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarRangePredicateExtractor.java (hunk 1 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarRangePredicateExtractor.java:129]: setRedundant seems only used for mv rewrite?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/JoinPredicatePushdown.java (hunk 3 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/JoinPredicatePushdown.java:410]: optimizerContext.addPushdownDerivedPredicate(p) is global effective, but isPushdown may be lost.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java (hunk 2 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java]: fixed.

**Review by satanson — DISMISSED** [review]

**Review by LiShuMing — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**